staging/provisioning/windows/windowssecuretls.ps1

```diff
@@ -0,0 +1,387 @@
+
+#******************* FUNCTION THAT ACTUALLY UPDATES KEYS; WILL RETURN REBOOT FLAG IF CHANGES ***********************
+function Set-CryptoSetting {
+    param ( 
+        $regKeyName, 
+        $value, 
+        $valuedata, 
+        $valuetype      
+    ) 
+
+    # Check for existence of registry key, and create if it does not exist 
+    If (!(Test-Path -Path $regKeyName)) { 
+        New-Item $regKeyName | Out-Null 
+    } 
+
+    # Get data of registry value, or null if it does not exist 
+    $val = (Get-ItemProperty -Path $regKeyName -Name $value -ErrorAction SilentlyContinue).$value 
+
+
+    If ($val -eq $null) { 
+        # Value does not exist - create and set to desired value 
+        New-ItemProperty -Path $regKeyName -Name $value -Value $valuedata -PropertyType $valuetype | Out-Null 
+    }
+    Else { 
+        # Value does exist - if not equal to desired value, change it 
+        If ($val -ne $valuedata) { 
+            Set-ItemProperty -Path $regKeyName -Name $value -Value $valuedata 
+        } 
+    } 
+}
+#***************************************************************************************************************
+
+#******************* FUNCTION THAT DISABLES RC4 ***********************
+function DisableRC4 {
+    $subkeys = Get-Item -Path "HKLM:\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL" 
+    $ciphers = $subkeys.OpenSubKey("Ciphers", $true) 
+
+    Write-Log "----- Checking the status of RC4 -----"
+
+    $RC4 = $false
+    if ($ciphers.SubKeyCount -eq 0) { 
+        $k1 = $ciphers.CreateSubKey("RC4 128/128") 
+        $k1.SetValue("Enabled", 0, [Microsoft.Win32.RegistryValueKind]::DWord) 
+        $k2 = $ciphers.CreateSubKey("RC4 64/128") 
+        $k2.SetValue("Enabled", 0, [Microsoft.Win32.RegistryValueKind]::DWord) 
+        $k3 = $ciphers.CreateSubKey("RC4 56/128") 
+        $k3.SetValue("Enabled", 0, [Microsoft.Win32.RegistryValueKind]::DWord) 
+        $k4 = $ciphers.CreateSubKey("RC4 40/128") 
+        $k4.SetValue("Enabled", 0, [Microsoft.Win32.RegistryValueKind]::DWord) 
+        
+        Write-Log "RC4 was disabled "
+        $RC4 = $true
+    } 
+
+    If ($RC4 -ne $true) {
+        Write-Log "There was no change for RC4 "
+    }
+}
+#***************************************************************************************************************
+
+#******************* FUNCTION CHECKS FOR PROBLEMATIC FIPS SETTING AND FIXES IT  ***********************
+function Test-RegistryValueForFipsSettings {            
+    $fipsPath = @( 
+        "HKLM:\System\CurrentControlSet\Control\Terminal Server\WinStations\RDP-Tcp",
+        "HKLM:\SOFTWARE\Policies\Microsoft\Windows NT\Terminal Services",
+        "HKLM:\System\CurrentControlSet\Control\Terminal Server\DefaultUserConfiguration"
+    )
+    
+    $fipsValue = "MinEncryptionLevel"
+    
+    foreach ($path in $fipsPath) {
+        Write-Log "Checking to see if $($path)\$fipsValue exists"

```

Comment:
Does **Write-Log** need to be defined or imported/dot-sourced in from another script file?

Reply:
Yes. This will be used in kuberneteswindowssetup.ps1 and Write-Log will be imported from parts/windows/kuberneteswindowsfunctions.ps1

Reply:
This follows the pattern in our CSE files, where the discrete script code file as-is is sort of vulnerable to not being "packaged" together with the dependent functions/vars. I think this is fine for the way that we maintain our scripts, and if we try to run this without proper linking the failure will be obvious/quick.